src/AuthRoles.js

```diff
@@ -0,0 +1,322 @@
+        // accept it if possible, we still need to check the acl !
+        if(this.isRoleAccessible(rolePreviouslyRejected)){
+          rolePreviouslyRejected.tag = OppResult.accepted;
+          addIfNeed(this.accessibleRoles.ids, rolePreviouslyRejected.objectId)
+          addIfNeed(this.accessibleRoles.names, "role:" + rolePreviouslyRejected.name)
+          // do the same of this role
+          this.resolvePreviousRejectionsIfPossible(rolePreviouslyRejected)
+        }else{
+          // role is still rejected, because its acl did not validate
+        }
+      }
+    })
+    // clear rejections, this is not required since we will never
+    // re-accept the same role twice, but freeing memory is never a bad idea.
+    delete this.rejections[ roleThatWasJustAccepted.objectId ]
+  }
+}
+
+
+// Returns :
+// inconclusive, rejected or accepted
+AuthRoles.prototype.isAnyPathValid = function(role){
+  const parentIds = role.parents
+  // assume rejected
+  var finalResult = OppResult.rejected
+  // compute each path individually
+  for (let index = 0; index < parentIds.length; index++) {
+    const parentId = parentIds[index];
+    const parentRole = this.manifest[parentId]
+    if(!parentRole) continue;
+
+    const pathResult = this.computeAccessOnRole(parentRole)
+    if(pathResult === OppResult.accepted){
+      // path accepted, skip all other paths and return
+      return OppResult.accepted
+    }else if(pathResult === OppResult.rejected){
+      // path rejected, but prioritize inconclusive over
+      // rejected.
+      if(finalResult !== OppResult.inconclusive){
+        finalResult = OppResult.rejected
+      }
+    }else if(pathResult === OppResult.inconclusive){
+      finalResult = OppResult.inconclusive
+    }
+
+    // mark that our 'role' has been rejected by 'parentRole'
+    if(pathResult !== OppResult.accepted){
+      this.markRejected(parentRole, role)
+    }
+  }
+  return finalResult
+}
+
+// A role is accessible when any of the following statements is valid :
+// 1- User is explicitly given access to the role
+// 2- Role has access to itself
+// 3- Role is accessible from other roles we have
+// 4- Role is publicly accessible
+AuthRoles.prototype.isRoleAccessible = function(role){
+
+  // Accept role regardless of its ACL if we are using Master.
+  // Usually when using Master, we will never reach here.
+  // This is basically only used for tests.
+  if(this.isMaster === true) return true;
+
+  const acl = role.ACL;
+  const userRoles = this.accessibleRoles.names
+
+  if(acl === {} || !acl){
+    return false
+  }
+  // (1)
+  if(isAnyExplicitlyGranted(acl, [this.userId])){
+    return true
+  }
+  // (2, 4)
+  if(isAnyExplicitlyGranted(acl, ["*", "role:" + role.name])){
+    return true
+  }
+  // (3)
+  if(isAnyExplicitlyGranted(acl, userRoles)){
+    return true
+  }
+  return false
+}
+
+// Or
+function isAnyExplicitlyGranted(acl, roleNames){
+  for (let index = 0; index < roleNames.length; index++) {
```

Comment:
use 

```
roleNames.find((name) => {
 /// logic here
});
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

Reply:
I have actually found that its faster to loop over the ACL and look through the role names from a set (.has())
I will change it for you to review